api/proquest/importer.py

```diff
@@ -835,48 +837,75 @@ def _get_publications(feed):
                     for publication in group.publications:
                         yield publication
 
-    def _download_feed_pages(self):
+    def _download_feed_pages(self, feed_pages_directory):
         """Download all the pages of the ProQuest OPDS feed.
 
         Downloaded feed pages are dumped to the local disk instead of storing them in memory
         in order to decrease the memory footprint.
 
-        :return: List of temporary file handles pointing to feed pages
-        :rtype: List[file]
-        """
-        feed_temporary_files = []
+        We have to wait until all the pages are downloaded because
+        the ProQuest feed structure doesn't allow us to use short-circuit logic
+        (there is no update date and ordering sometimes is incorrect),
+        and we always have to start from scratch.
+        Hence, we need to have all the pages locally before starting the import process.
+        Otherwise, the import won't succeed.
+
+        :param feed_pages_directory: Absolute path to the directory containing temporary files with feed pages
+        :type feed_pages_directory: str
 
+        :return: List of absolute file names containing feed pages
+        :rtype: List[str]
+        """
         self._logger.info("Started downloading feed pages")
 
+        feed_page_files = []
+
         for feed in self._client.download_all_feed_pages(self._db):
-            feed_temporary_file = tempfile.TemporaryFile(mode="r+")
-            feed_temporary_file.write(json.dumps(feed))
-            feed_temporary_file.flush()
+            feed_page_file = tempfile.NamedTemporaryFile(
+                mode="r+", dir=feed_pages_directory, delete=False
+            )
 
-            feed_temporary_files.append(feed_temporary_file)
+            try:
+                feed_page_content = json.dumps(
+                    feed, default=str, ensure_ascii=True, encoding="utf-8"
+                )
+                feed_page_file.write(feed_page_content)
+                feed_page_file.flush()
+
+                feed_page_files.append(feed_page_file.name)
+            except:
+                self._logger.exception(
+                    "An unexpected error occurred during saving a feed page"
+                )
+
+                feed_page_file.delete()
```

Comment:
If we're on the second or subsequent feed page, will all previous pages get deleted?

Might be easier to let `ProQuestOPDS2ImportMonitor._parse_feed_page` and handle clean up of the entire `feed_pages_directory`.

Reply:
I agree; the code that makes the temporary directory should be in charge of clearing out anything in that directory before deleting it.

Reply:
This part is actually not required, since there is `raise` statement the stack will be unwind and got to the `finally` block in `_get_feeds` method where `shutil.rmtree` will take care of recursively removing the temporary directory with all its content.

Reply:
Yes.

Reply:
I'm OK merging this either with `feed_page_file.delete` or without it, though I'd prefer it be removed. LMK what you'd prefer.

Reply:
I would prefer that `feed_page_file.delete` be removed, since it might confuse someone later. That someone might be me. 🙃

Reply:
I removed `feed_page_file.delete` and extended the log message a bit but anyways it should contain all the required information about the exception because we use `exception` method.